src/lib/operators/window_function_evaluator.cpp

```diff
@@ -0,0 +1,607 @@
+#include "hyrise.hpp"
+#include "resolve_type.hpp"
+#include "scheduler/job_task.hpp"
+#include "storage/segment_iterate.hpp"
+#include "storage/value_segment.hpp"
+#include "types.hpp"
+#include "utils/assert.hpp"
+#include "utils/format_duration.hpp"
+#include "utils/segment_tree.hpp"
+#include "utils/timer.hpp"
+#include "window_function_evaluator.hpp"
+
+namespace hyrise::window_function_evaluator {
+
+WindowFunctionEvaluator::WindowFunctionEvaluator(
+    const std::shared_ptr<const AbstractOperator>& input_operator, std::vector<ColumnID> init_partition_by_column_ids,
+    std::vector<ColumnID> init_order_by_column_ids, ColumnID init_function_argument_column_id,
+    std::shared_ptr<WindowFunctionExpression> init_window_funtion_expression)
+    : AbstractReadOnlyOperator(OperatorType::WindowFunction, input_operator, nullptr,
+                               std::make_unique<PerformanceData>()),
+      _partition_by_column_ids(std::move(init_partition_by_column_ids)),
+      _order_by_column_ids(std::move(init_order_by_column_ids)),
+      _function_argument_column_id(init_function_argument_column_id),
+      _window_function_expression(std::move(init_window_funtion_expression)) {
+  Assert(
+      _function_argument_column_id != INVALID_COLUMN_ID || is_rank_like(_window_function_expression->window_function),
+      "Could not extract window function argument, although it was not rank-like");
+}
+
+const std::string& WindowFunctionEvaluator::name() const {
+  static const auto name = std::string{"WindowFunctionEvaluator"};
+  return name;
+}
+
+const FrameDescription& WindowFunctionEvaluator::frame_description() const {
+  const auto window = std::dynamic_pointer_cast<WindowExpression>(_window_function_expression->window());
+  return window->frame_description;
+}
+
+void WindowFunctionEvaluator::_on_set_parameters(
+    [[maybe_unused]] const std::unordered_map<ParameterID, AllTypeVariant>& parameters) {}
+
+std::shared_ptr<AbstractOperator> WindowFunctionEvaluator::_on_deep_copy(
+    const std::shared_ptr<AbstractOperator>& copied_left_input,
+    [[maybe_unused]] const std::shared_ptr<AbstractOperator>& copied_right_input,
+    [[maybe_unused]] std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>>& copied_ops) const {
+  return std::make_shared<WindowFunctionEvaluator>(copied_left_input, _partition_by_column_ids, _order_by_column_ids,
+                                                   _function_argument_column_id, _window_function_expression);
+}
+
+template <typename InputColumnType, WindowFunction window_function>
+std::shared_ptr<const Table> WindowFunctionEvaluator::_templated_on_execute() {
+  if (!is_rank_like(window_function) && frame_description().type == FrameType::Range) {
+    Assert(_order_by_column_ids.size() <= 1,
+           "For non-rank-like window functions, range mode frames are only allowed when there is at most one order-by "
+           "expression.");
```

Comment:
It says so in the paper :D It also makes sense, because for 0 order-by columns, all tuples are equivalent and with 1 column you can have a reasonable sense of "range" - with two columns, it's not so clear anymore. Can elaborate in a comment though